// react-scrollify/dist/components/VerticalRail.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _helpers = require('../helpers');

var _helpers2 = _interopRequireDefault(_helpers);

var _VerticalBar = require('./VerticalBar');

var _VerticalBar2 = _interopRequireDefault(_VerticalBar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var VerticalRail = function (_React$Component) {
    (0, _inherits3.default)(VerticalRail, _React$Component);

    function VerticalRail(props) {
        (0, _classCallCheck3.default)(this, VerticalRail);

        var _this = (0, _possibleConstructorReturn3.default)(this, (VerticalRail.__proto__ || (0, _getPrototypeOf2.default)(VerticalRail)).call(this, props));

        _this.handleOnClick = _this.handleOnClick.bind(_this);
        _this.updateScrollAndGeometry = _this.updateScrollAndGeometry.bind(_this);
        return _this;
    }

    (0, _createClass3.default)(VerticalRail, [{
        key: 'componentDidMount',
        value: function componentDidMount() {}
    }, {
        key: 'handleOnClick',
        value: function handleOnClick(e) {
            if (e.target !== _reactDom2.default.findDOMNode(this.refs.ybar)) {
                var _props = this.props,
                    verticalBarTop = _props.verticalBarTop,
                    containerHeight = _props.containerHeight;

                var positionTop = e.pageY - window.pageYOffset - _reactDom2.default.findDOMNode(this).getBoundingClientRect().top;
                var direction = positionTop > verticalBarTop ? 1 : -1;
                var newTop = _reactDom2.default.findDOMNode(this).parentNode.scrollTop + direction * containerHeight;
                this.updateScrollAndGeometry(newTop);
            }

            e.stopPropagation();
        }
    }, {
        key: 'updateScrollAndGeometry',
        value: function updateScrollAndGeometry(newTop) {
            // this.props.updateContainerScroll('top', newTop);
            // this.props.updateGeometry();
            // this.props.updateState();
        }
    }, {
        key: 'render',
        value: function render() {
            var style = {
                height: this.props.verticalRailHeight + 'px',
                top: this.props.verticalRailTop + 'px'
            };
            if (this.props.isVerticalBarUsingRight) {
                style['right'] = this.props.verticalRailRight + 'px';
            } else {
                style['left'] = this.props.verticalRailLeft + 'px';
            }

            return _react2.default.createElement(
                'div',
                {
                    className: 'ps-scrollbar-y-rail',
                    style: style,
                    onClick: this.handleOnClick
                },
                _react2.default.createElement(_VerticalBar2.default, this.props)
            );
        }
    }]);
    return VerticalRail;
}(_react2.default.Component);

exports.default = VerticalRail;
